[Bug report] Built-in job template run fails with "Source file does not exist" after Gravitino version upgrade

### Version

main branch

### Describe what's wrong

After upgrading Gravitino across versions (e.g. via a rolling update on Kubernetes), running a built-in job template (e.g. `builtin-iceberg-rewrite-data-files`) fails because the server still tries to load the old-version jar path (`gravitino-jobs-<old-version>.jar`) that no longer exists in the container.

The root cause is in `BuiltInJobTemplateEventListener.reconcileBuiltInJobTemplates()` (`core/src/main/java/org/apache/gravitino/job/BuiltInJobTemplateEventListener.java`) — it only compares the template's `version` custom field (e.g. `"v1"`) to decide whether to refresh the persisted record. Since the built-in job's logic didn't change, `version` stays at `"v1"` across a Gravitino release bump, so the `executable` path — which is resolved dynamically at startup from the on-disk jar location via `BuiltInJob.resolveExecutable()` (`maintenance/jobs/src/main/java/org/apache/gravitino/maintenance/jobs/BuiltInJob.java`) — is never refreshed in the database, even though the actual jar filename changed with the release.

### Error message and/or stacktrace

```
{
  "code": 1002,
  "type": "RuntimeException",
  "message": "Failed to operate object operation [RUN] under [test], reason [Failed to fetch file from URI /opt/gravitino/auxlib/gravitino-jobs-1.4.0-SNAPSHOT.jar]",
  "stack": [
    "java.lang.RuntimeException: Failed to fetch file from URI /opt/gravitino/auxlib/gravitino-jobs-1.4.0-SNAPSHOT.jar",
    "at org.apache.gravitino.job.JobManager.fetchFileFromUri(JobManager.java:812)",
    "at org.apache.gravitino.job.JobManager.createRuntimeJobTemplate(JobManager.java:677)",
    "at org.apache.gravitino.job.JobManager.runJob(JobManager.java:438)",
    "...",
    "Caused by: java.io.IOException: Source file does not exist: /opt/gravitino/auxlib/gravitino-jobs-1.4.0-SNAPSHOT.jar",
    "at org.apache.gravitino.utils.FileFetcher.linkLocalFile(FileFetcher.java:129)",
    "at org.apache.gravitino.utils.FileFetcher.fetchFileFromUri(FileFetcher.java:107)",
    "at org.apache.gravitino.job.JobManager.fetchFileFromUri(JobManager.java:806)"
  ]
}
```

Container actual state after upgrade:

```
$ ls /opt/gravitino/auxlib/
gravitino-cli-2.0.0-SNAPSHOT.jar  gravitino-jobs-2.0.0-SNAPSHOT.jar
```

### How to reproduce

1. Deploy a Gravitino build with built-in job templates enabled (e.g. via Helm chart on Kubernetes)
2. Run any built-in job (e.g. `builtin-iceberg-rewrite-data-files`) — this succeeds and persists the template to the database with `executable = /opt/gravitino/auxlib/gravitino-jobs-<old-version>.jar`
3. Upgrade the Gravitino image to a new version (rolling update), which ships `gravitino-jobs-<new-version>.jar`
4. Run the same built-in job again — fails with the error above

### Additional context

`reconcileBuiltInJobTemplates()` should also detect changes to the `executable` path (and other mutable fields like arguments/configs), not rely solely on the `version` string, when deciding whether to refresh the persisted template. `JobTemplate` already implements a full `equals()`, so the fix can compare `existing.toJobTemplate().equals(newTemplate)` in addition to the version check.

--- a/core/src/main/java/org/apache/gravitino/job/BuiltInJobTemplateEventListener.java
+++ b/core/src/main/java/org/apache/gravitino/job/BuiltInJobTemplateEventListener.java
@@ -269,7 +269,12 @@ public class BuiltInJobTemplateEventListener implements EventListenerPlugin {
           } else {
             int existingVersion = version(existing.templateContent().customFields());
             int newVersion = version(newTemplate.customFields());
-            if (newVersion > existingVersion) {
+            // The version field only tracks changes to job logic, but fields like `executable`
+            // can drift independently (e.g. the jar path changes with the Gravitino release
+            // version even though the job logic version stays the same), so also refresh the
+            // persisted template whenever its content no longer matches, regardless of version.
+            boolean contentChanged = !existing.toJobTemplate().equals(newTemplate);
+            if (newVersion > existingVersion || contentChanged) {
               updateBuiltInJobTemplate(metalake, existing, newTemplate);
             } else {
               LOG.info("Built-in job template {} under metalake {} is up to date", name, metalake);

--- a/core/src/test/java/org/apache/gravitino/job/TestBuiltInJobTemplateEventListener.java
+++ b/core/src/test/java/org/apache/gravitino/job/TestBuiltInJobTemplateEventListener.java
@@ -416,10 +416,11 @@ public class TestBuiltInJobTemplateEventListener {
     String metalakeName = "test_metalake";
     Map<String, JobTemplate> builtInTemplates = new HashMap<>();
 
+    // Same version and identical content as the entity created by createJobTemplateEntity(...)
     ShellJobTemplate template =
         ShellJobTemplate.builder()
             .withName("builtin-existing")
-            .withComment("same version")
+            .withComment("test")
             .withExecutable("/bin/echo")
             .withCustomFields(Collections.singletonMap("version", "v1"))
             .build();
@@ -444,6 +445,44 @@ public class TestBuiltInJobTemplateEventListener {
             Entity.EntityType.JOB_TEMPLATE,
             JobTemplateEntity.class);
     Assertions.assertEquals("v1", result.templateContent().customFields().get("version"));
+  }
+
+  @Test
+  public void testReconcileBuiltInJobTemplatesUpdateWhenExecutableChangedButVersionSame()
+      throws IOException {
+    String metalakeName = "test_metalake";
+    Map<String, JobTemplate> builtInTemplates = new HashMap<>();
+
+    // Same version as the persisted entity, but a different executable path, e.g. because the
+    // Gravitino release version (and thus the jar filename) changed while the job logic did not.
+    ShellJobTemplate template =
+        ShellJobTemplate.builder()
+            .withName("builtin-existing")
+            .withComment("test")
+            .withExecutable("/opt/gravitino/auxlib/gravitino-jobs-2.0.0-SNAPSHOT.jar")
+            .withCustomFields(Collections.singletonMap("version", "v1"))
+            .build();
+    builtInTemplates.put("builtin-existing", template);
+
+    // Existing entity was persisted with the old jar path.
+    JobTemplateEntity existingEntity = createJobTemplateEntity("builtin-existing", "v1");
+    entityStore.put(existingEntity, false);
+
+    when(jobManager.listJobTemplates(metalakeName))
+        .thenReturn(Collections.singletonList(existingEntity));
+
+    listener.reconcileBuiltInJobTemplates(metalakeName, builtInTemplates);
+
+    // Verify the entity's executable path was refreshed despite the version staying at v1
+    JobTemplateEntity result =
+        entityStore.get(
+            existingEntity.nameIdentifier(),
+            Entity.EntityType.JOB_TEMPLATE,
+            JobTemplateEntity.class);
+    Assertions.assertEquals("v1", result.templateContent().customFields().get("version"));
+    Assertions.assertEquals(
+        "/opt/gravitino/auxlib/gravitino-jobs-2.0.0-SNAPSHOT.jar",
+        result.templateContent().executable());
   }
 
   @Test

--- a/core/src/test/java/org/apache/gravitino/job/TestBuiltInJobTemplateEventListener.java
+++ b/core/src/test/java/org/apache/gravitino/job/TestBuiltInJobTemplateEventListener.java
@@ -445,6 +445,12 @@ public class TestBuiltInJobTemplateEventListener {
             Entity.EntityType.JOB_TEMPLATE,
             JobTemplateEntity.class);
     Assertions.assertEquals("v1", result.templateContent().customFields().get("version"));
+
+    // Verify no update actually happened: an update would stamp lastModifier/lastModifiedTime,
+    // which createJobTemplateEntity(...) leaves null, so these staying null proves the entity
+    // was left untouched rather than merely ending up with the same version by coincidence.
+    Assertions.assertNull(result.auditInfo().lastModifier());
+    Assertions.assertNull(result.auditInfo().lastModifiedTime());
   }
 
   @Test
@@ -591,6 +597,11 @@ public class TestBuiltInJobTemplateEventListener {
         .withId(1L)
         .withName(name)
         .withNamespace(NamespaceUtil.ofJobTemplate("test_metalake"))
+        // TemplateContent does not carry comment (see JobTemplateEntity#toJobTemplate(), which
+        // reads comment from the entity itself), so it must be set explicitly here to match what
+        // registerNewBuiltInJobTemplate(...) does in production and keep entity.toJobTemplate()
+        // consistent with `template` above.
+        .withComment(template.comment())
         .withTemplateContent(JobTemplateEntity.TemplateContent.fromJobTemplate(template))
         .withAuditInfo(
             AuditInfo.builder().withCreator("test").withCreateTime(Instant.now()).build())

